core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -82,48 +100,157 @@ public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardR
+            inSyncAllocationIds.removeAll(updates.removedAllocationIds);
+
+            // Prevent set of inSyncAllocationIds to grow unboundedly. This can happen for example if we don't write to a primary
+            // but repeatedly shut down nodes that have active replicas.
+            // We use number_of_replicas + 1 (= possible active shard copies) to bound the inSyncAllocationIds set
+            int upperBound = oldIndexMetaData.getNumberOfReplicas() + 1; // +1 for the primary
+            int overflow = inSyncAllocationIds.size() - upperBound;
+            if (overflow > 0) {
+                // trim entries that have no corresponding shard routing in the cluster state (i.e. trim unavailable copies)
+                Set<String> assignedAllocations = newRoutingTable.shardRoutingTable(shardId).assignedShards().stream()
+                    .map(shr -> shr.allocationId().getId()).collect(Collectors.toSet());
+                Set<String> candidatesToRemove = Sets.difference(inSyncAllocationIds, assignedAllocations);
+                for (Iterator<String> it = candidatesToRemove.iterator(); it.hasNext() && overflow > 0; ) {
+                    inSyncAllocationIds.remove(it.next());
+                    overflow--;
+                }
+                assert overflow == 0 :
+                    "not enough candidates to remove, candidates: " + candidatesToRemove + " overflow: " + overflow;
+            }
+
+            // only update in-sync allocation ids if there is at least one entry remaining. Assume for example that there only
+            // ever was a primary active and now it failed. If we were to remove the allocation id from the in-sync set, this would
+            // create an empty primary on the next allocation (see ShardRouting#allocatedPostIndexCreate)
+            if (inSyncAllocationIds.isEmpty() && oldInSyncAllocationIds.isEmpty() == false) {
+                assert updates.firstFailedPrimary != null :
+                    "in-sync set became empty but active primary wasn't failed: " + oldInSyncAllocationIds;
+                if (updates.firstFailedPrimary != null) {
+                    // add back allocation id of failed primary
+                    inSyncAllocationIds.add(updates.firstFailedPrimary.allocationId().getId());
+                }
+            }
+
+            assert inSyncAllocationIds.isEmpty() == false || oldInSyncAllocationIds.isEmpty() :
+                "in-sync allocations cannot become empty after they have been non-empty: " + oldInSyncAllocationIds;
+
+            // be extra safe here and only update in-sync set if it is non-empty
+            if (inSyncAllocationIds.isEmpty() == false) {
+                if (indexMetaDataBuilder == null) {
+                    indexMetaDataBuilder = IndexMetaData.builder(oldIndexMetaData);
                 }
+                indexMetaDataBuilder.putInSyncAllocationIds(shardId.id(), inSyncAllocationIds);
+            }
```

Comment:
yeah, I thought about that together with @abeyad. The only interesting thing  about logging this is to see that "this unexpected thing happened". Logging it at trace/debug level means that we never get to see it unless we're actively looking for it (then again I'm not sure in which scenarios I would ever think about turning the log level of this class to trace to detect this as being the cause for some other issues). I've opted to use an assertion instead so that we at least have a hard failure in our tests. If we want to log this, it probably makes more sense to do it at WARN level so that we get active feedback from users if this happens. I've been going back and forth with this. Any additional thoughts @dakrone and @abeyad?